Sort option w/FIFO default and tests, Priority fix

niftylettuce opened the pull request:
This closes #462 and #440

OmgImAlexis commented on .eslintrc:
```diff
@@ -15,8 +15,8 @@ rules:
-    space-after-function-name: [2, "never"]
-    space-after-keywords: [2, "always"]
+    space-before-function-paren: [2, "never"]
+    keyword-spacing: [2, { "after": true }]
```
Please remove the changes to the eslint.

niftylettuce commented:
The current one is deprecated, I fixed deprecations

On Jul 15, 2017 5:34 AM, "Alexis Tyler" <notifications@github.com> wrote:

> *@OmgImAlexis* requested changes on this pull request.
> ------------------------------
>
> In .eslintrc
> <https://github.com/agenda/agenda/pull/464#discussion_r127580827>:
>
> > @@ -15,8 +15,8 @@ rules:
>      new-cap: 0
>      radix: 2
>      quotes: 0
> -    space-after-function-name: [2, "never"]
> -    space-after-keywords: [2, "always"]
> +    space-before-function-paren: [2, "never"]
> +    keyword-spacing: [2, { "after": true }]
>
> Please remove the changes to the eslint.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/agenda/agenda/pull/464#pullrequestreview-50185932>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAf7heYWPc8C0SSzasZUHBICoiBqpU8Xks5sOIejgaJpZM4OYqp1>
> .
>

OmgImAlexis commented:
This PR is for adding first in first out, not updating rules. PRs should be topic specific so if we have to revert, etc. we don't have to cherry pick the commits.

If you want to update the eslint rules please do it in another PR.

niftylettuce commented:
Why all the extra work for something so simple? You're defeating the
purpose of contributions and shipping updates vs having some illusion of
organization. But okay..

On Jul 15, 2017 12:07 PM, "Alexis Tyler" <notifications@github.com> wrote:

> This PR is for adding first in first out, not updating rules. PRs should
> be topic specific so if we have to revert, etc. we don't have to cherry
> pick the commits.
>
> If you want to update the eslint rules please do it in another PR.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/agenda/agenda/pull/464#issuecomment-315544167>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAf7haYq-zjo8UWFHtzSQoLxCh7SRllKks5sOOO7gaJpZM4OYqp1>
> .
>

niftylettuce commented:
Added support for custom sorting as requested @OmgImAlexis 

The codebase in `lib/agenda.js` is a mess, there's mixed spaces/indentation and it doesn't seem like anyone even adhered to the eslint rules anyways 😦

OmgImAlexis commented:
> The codebase in `lib/agenda.js` is a mess, there's mixed spaces/indentation and it doesn't seem like anyone even adhered to the eslint rules anyways.

Working on this in https://github.com/agenda/agenda/pull/465

simison commented:
>Why all the extra work for something so simple? You're defeating the purpose of contributions and shipping updates vs having some illusion of organization.

No stress, it just safeguards getting some work in even if some PR ends up being long-running (as they sometimes tend to be at Agenda). ;-) Also adds less noise for reviewing.

Thanks a bunch for PR! Awesome stuff.

OmgImAlexis commented:
Closing and reopening to allow travis to re-run so coveralls can pass. I'll merge this after that if no-one objects.

lushc commented:
Any chance of adding a test for the scenario described in the FAQ? Right now we're missing any kind of coverage regarding job running order. The test in #451 could be adapted to include a 2nd job with the same priority.

OmgImAlexis commented:
If @niftylettuce doesn't get to this by tomorrow midday (GMT+1030) I'll add it.

niftylettuce commented:
@OmgImAlexis I went ahead and removed eslint changes and rebased, but getting odd conflict I'm not sure about?

OmgImAlexis commented:
@niftylettuce did you want to add the test from #451?

niftylettuce commented:
@OmgImAlexis way too tired, just spent 10 mins at least not realizing I had to merge w/upstream haha.... 😆

OmgImAlexis commented:
All good, I'll add it as soon as I'm free.

OmgImAlexis commented:
After spending 10 minutes trying to workout why I couldn't pull this branch down I realised it's in @niftylettuce's fork. Could you please add the tests over as I can't push to your account.

If you can't and/or don't have time we'll merge and I'll add them to master via another PR.

simison commented:
@niftylettuce please tick "[ ] Allow edits from maintainers" ([instructions](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/))

![image](https://user-images.githubusercontent.com/87168/28368061-c7517d6e-6c9b-11e7-89b0-989323b09cb6.png)

niftylettuce commented:
<img width="201" alt="screen shot 2017-07-19 at 12 51 18 pm" src="https://user-images.githubusercontent.com/523141/28379236-ffa3eb72-6c80-11e7-8c9c-2e1c45daf67a.png">

It already was ticked?

niftylettuce commented:
Tested are added, this closes #462 and #440

niftylettuce commented:
Not sure why Travis failed only for v6.x - it literally just passed.  Maybe we need to use `jobTimeout` in the tests?  Let me know @OmgImAlexis

lushc commented on lib/job.js:
```diff
@@ -14,7 +14,13 @@ var Job = module.exports = function Job(args) {
   // Process args
-  args.priority = parsePriority(args.priority) || 0;
+
+  // allow priority overrides (e.g. `job.now('foo', { priority: 'low' })`)
```
Unfortunately I don't agree with this change as there's already a [method to change priority](https://github.com/agenda/agenda#prioritypriority) in the job API. Introducing the concept of reserved words and having Agenda's behaviour leak into the data object will just cause problems IMO. For example, breaking someone's job because their domain data also has a priority key.

niftylettuce replied:
Do you have any thoughts on how best to rewrite my tests to for priority and FIFO?  That was my main use case (besides the issue on GH) for adding this.

lushc replied:
Ah right I see, I would use the job API directly and cascade the saves, similar to how to how it was done in https://github.com/agenda/agenda/pull/451/commits/336ff6445803606a6dc468a6f26c637145790adc (please forgive the variable names):

```js
var now = Date.now();
var one = jobs.create('fifo-priority').schedule(now).priority('high');
var two = jobs.create('fifo-priority').schedule(now + 100).priority('low');
var three = jobs.create('fifo-priority').schedule(now + 200).priority('high');

one.save(function () {
  two.save(function () {
    three.save(function () {
      jobs.start();
    });
  });
});
// check jobs ...
```

niftylettuce replied:
@lushc Ah, be careful - you needed to wrap `schedule(now)` and `schedule(now + 200)` with `schedule(new Date(now))` and `schedule(new Date(now + 200))`.  Agenda by default doesn't parse a timestamp as a date. 

This seems to be an actual bug - shouldn't we support passing a timestamp from `Date.now()` or `new Date().getTime()`? cc @OmgImAlexis

OmgImAlexis replied:
Yep found this out while debugging some of our tests.

Open an issue? Since we don't have loads of docs on what should or shouldn't be supported I'd say it's a bug.

lushc commented on lib/agenda.js:
```diff
@@ -24,6 +24,7 @@ var Agenda = module.exports = function(config, cb) {
   this._lockedJobs = [];
   this._jobQueue = [];
   this._defaultLockLifetime = config.defaultLockLifetime || 10 * 60 * 1000; // 10 minute default lockLifetime
+  this._sort = config.sort || { priority: -1, _id: 1 };
```
The indexes created in `db_init` and `handleLegacyCreateIndex` will need updating to account for this. I'd merge the sort config in at the `priority` key's current position.

niftylettuce replied:
I'm not sure what you mean by this - do you mean just bump it up a few lines?

lushc replied:
Apologies for not being clear enough, I typed that on my phone so I gave up trying to give an example!

[Indexes are automatically created](https://github.com/agenda/agenda/blob/7c4692a1c4586d6078ccd03538c6f6b22896722f/lib/agenda.js#L103) and the order is quite important so Mongo can take advantage of it during sorting. So we should be adding the sort config in the right place, e.g.

```js
this._collection.createIndexes([{
  "key": Object.assign({"name": 1}, config.sort, {"lockedAt": 1, "nextRunAt": 1, "disabled": 1}),
  "name": "findAndLockNextJobIndex1"
} // ...
```

niftylettuce replied:
@lushc Why do we need two indices by the way?

simison commented:
>Not sure why Travis failed only for v6.x - it literally just passed.

@OmgImAlexis and @lushc observed tests failing kinda randomly earlier as well. I could find only [this comment](https://github.com/agenda/agenda/pull/476#issuecomment-316730252) referring to it though.

simison commented:
FYI moved this to 1.0.0 milestone as [discussed at the issue](https://github.com/agenda/agenda/issues/462#issuecomment-316928431).

simison commented:
@niftylettuce it's a bit hard now to see what's going on with all the eslint changes to test file. :-) Could you move them to a separate PR and we can just quickly get it merged? Sorry for the hassle but it's just to make everyone's day easier. <3

Imagine when people who are updating their Agenda and want to observe/study this change later on via commit logs.

smranta commented:
@niftylettuce @simison sort by _id might not work for recurring jobs. Why dont we use nextRunAt instead ?

niftylettuce commented:
@smranta such a good suggestion, `nextRunAt` would be better.

simison commented:
We're waiting for npm rights to publish 0.10.0 and this one can go in next up.

@OmgImAlexis can you rebase this as you know how to modify those test files? You can clone https://github.com/niftylettuce/agenda and you should be able to push to that master branch [_because_](https://github.com/agenda/agenda/pull/464#issuecomment-316448606).

simison commented:
Someone also needs to address changes @lushc requested.

OmgImAlexis commented on lib/job.js:
```diff
@@ -14,7 +14,13 @@ var Job = module.exports = function Job(args) {
+  if (args.data && typeof args.data === 'object' && typeof args.data.priority !== 'undefined')
+    args.priority = parsePriority(args.data.priority) || 0;
+  else
+    args.priority = parsePriority(args.priority) || 0;
```
All `if` and `else` should use `{}` we should not be relying on indentation at all.

OmgImAlexis commented on package.json:
```diff
@@ -1,29 +1,8 @@
-    "scheduler",
-    "runner"
-  ],
+  "version": "0.9.1",
```
I'd suggest fixing up the `package.json` so the reordering of fields doesn't get merged as it's not related to this PR.

OmgImAlexis commented:
@simison any news on pushing 0.10.0 to npm?

Ideally I want to get my PR for the tests merged before this so this PR doesn't show changes in the test files due to whitespace, etc.

niftylettuce commented:
I will be going through all comments here, rebasing, and updating per requests soon.  Just need a bit more time.  Thank you,

niftylettuce commented:
@lushc @simison @OmgImAlexis please see <https://github.com/agenda/agenda/pull/464/commits/ee6c00bb716f4a077d9691df99a1475c30def7cc> - I've update the pull requests, rewritten tests, excluded package.json changes, excluded massive changes to test files, rebased / merged with upstream master, and it should be OK.  Can you all have a look at it?  I slightly changed the indexing stuff because I wasn't sure why we'd have to indices to begin with?

niftylettuce commented:
@OmgImAlexis I've made requested changes and pushed again, please review once more, thank you.

niftylettuce commented:
@OmgImAlexis I just fixed CodeClimate issue (duplicate code), should be OK now.

niftylettuce commented:
@OmgImAlexis Ah, just saw there was one place that I forgot to bump `checkResults` function up in the tests - that's good now... Hopefully that should be it!!!!

niftylettuce commented:
@OmgImAlexis I also opened an issue for the milliseconds issue per <https://github.com/agenda/agenda/issues/496>

niftylettuce commented:
@simison it looks like Travis failed on two counts - due to a timeout of 8000ms+?  not sure why, everything else passed.

smranta commented on README.md:
```diff
@@ -732,6 +738,15 @@ agenda.on('fail:send email', function(err, job) {
 
+### What is the order in which jobs run?
+
+Jobs are run **with priority in a first in first out order** (so they will be run in the order they were queued/created _AND_ with respect to highest priority).  We utilize MongoDB's ObjectID through the job record property of `_id` as a sorting value (since it is a [timestamp](https://docs.mongodb.com/manual/reference/method/ObjectId.getTimestamp/) in itself).
```
@niftylettuce This is conflicting text, since we are using nextRunAt and not _id.

Suggested text : 
Jobs are run **with priority in a first scheduled first out order** (so they will be run in the order they were scheduled _AND_ with respect to highest priority)

niftylettuce commented:
@smranta I've gone and pushed a new version with the Readme fixed per your suggestion.  Thank you.

OmgImAlexis commented on README.md:
```diff
@@ -291,6 +291,12 @@ You can also specify it during instantiation
+
+Takes a `query` which specifies the sort query to be used for finding and locking the next job.
+
+By default it is `{ nextRunAt: 1, priority: -1 }`, which obeys a first in first out approach, with respect to priority.
```
Shouldn't this be `{_id: 1, priority: -1}`?

OmgImAlexis replied:
Ah noticed you changed the other to the same. All good.

niftylettuce commented:
Thanks @OmgImAlexis - do we need another set of eyes on this or we good to merge it?

OmgImAlexis commented:
I want to merge but I'm not 100% sure on how we're going to cut the last npm version. This should be able to be merged into master and then we can just cut [0.10.0](https://github.com/agenda/agenda/milestone/1) from an old commit.

PING: @simison @lushc

simison commented:
For 0.10.0 we had some bugfixing etc going in, right? So would be good to get that in before breaking release so that those who are not ready to do big upgrade, could get at least smaller fixes in.

But if this starts to lag too many days I'd just merge in order to be able to work, and skip 0.10.0.

I'll comment on status of npm rights a #441

OmgImAlexis commented:
@simison we should be able to checkout https://github.com/agenda/agenda/commit/e239b6eed861ba9f053f944881192fb38a25052a from master and publish that as `0.10.0` to npm.

That would then allow this to be merged into master as the first of `1.0.0`.

OmgImAlexis commented:
Since `v0.10.0` has been pushed and we have no issues I'm merging this.

niftylettuce commented:
Can we release this to NPM @OmgImAlexis? 🎉

OmgImAlexis commented:
Do we have all the documentation setup? @simison maybe we move the other `1.0.0` changes to the next milestone?

simison commented:
Sure! Small releases are nice.

Just ensure first docs are there and explain well at changelog how this is
a breaking change.

On 8 Aug 2017 08:37, "Alexis Tyler" <notifications@github.com> wrote:

Do we have all the documentation setup? @simison
<https://github.com/simison> maybe we move the other 1.0.0 changes to the
next milestone?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/agenda/agenda/pull/464#issuecomment-320855402>, or mute
the thread
<https://github.com/notifications/unsubscribe-auth/AAFUgD_6NxHoenYjX04g4Hnhntwqzo4Rks5sV_QegaJpZM4OYqp1>
.

niftylettuce commented:
I believe it's good to go!

On Aug 8, 2017 1:37 AM, "Alexis Tyler" <notifications@github.com> wrote:

> Do we have all the documentation setup? @simison
> <https://github.com/simison> maybe we move the other 1.0.0 changes to the
> next milestone?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/agenda/agenda/pull/464#issuecomment-320855402>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAf7hZMHJ3VqQCOh79U57h_Q4cH0OF3wks5sV_QfgaJpZM4OYqp1>
> .
>

OmgImAlexis commented:
If you both are happy with this I'll publish master as is with just the FIFO fix as `1.0.0`. We can then start on `1.1.0`.

OmgImAlexis commented:
Looks like we already added docs stating https://github.com/agenda/agenda/pull/475 would be in `1.0.0`.

I'll fix the travis issue so that can be merged as well.

@niftylettuce I think we should add something [here](https://github.com/agenda/agenda#known-issues) about it.

simison commented:
@OmgImAlexis Also keeping https://github.com/agenda/agenda/blob/master/History.md is important — that's where people head to when updating.

OmgImAlexis commented:
Didn't realise we had a history.md I usually use the [release page](https://github.com/agenda/agenda/releases/tag/v0.10.0) to document that.

simison commented:
Yeah, keeping infos at either place works. It's easier for other people to help with keeping it up to date when it's in a file, though.

OmgImAlexis commented:
@niftylettuce could you add that since you seem to be better at writing up explanations than myself.

niftylettuce commented:
@OmgImAlexis I believe that @simison already added it <https://github.com/agenda/agenda/blob/master/History.md#breaking>.  Thank you @simison ❤️
